### ISSUE TITLE
Download files created during code execution

### ISSUE THREAD

**Issue by EtiennePerot** [EtiennePerot/open-webui-code-execution#4]
Right now, if code writes files inside the sandbox, they are simply lost forever. It would be very cool to be able to run code that generates files and then have these files be downloadable from the chat UI.

This idea came from a comment by @sultanjulyan in [this issue](https://github.com/open-webui/open-webui/issues/1321#issuecomment-2332403456).

**Comment by sultanjulyan** [comment]
That's right, I want to create a function that can generate a file and then download it directly from the web UI. Thank you for your attention, I will wait for further developments.

**Comment by EtiennePerot** [comment]
From some preliminary research, it seems tools and functions can append messages to Open WebUI, but they are limited to either text (markdown) an images. So something like a spreadsheet file (`.xlsx`) would not be able to be displayed, unless perhaps the tool also had code to convert the spreadsheet to a Markdown table.

Alternatively, it may be possible to show links to download files that aren't markdown/images. Need to look if Open WebUI has some passthrough endpoints that would allow this, or if they could be linked using `data` URIs (for small files like spreadsheets this might work too).

**Comment by sultanjulyan** [comment]
Are there any alternative libraries that allow for that besides Web UI?

**Comment by EtiennePerot** [comment]
Not really. At the end of the day, the output generated by the tool need to be displayed on Open WebUI (that's simply the nature of this tool). We have no control over what functionality Open WebUI will display, but it may be possible to send them PRs to expand this.

Anyway, first feature to implement here is to detect the presence of files at the end of a code execution run and to copy them somewhere outside of the sandbox before the sandbox disappears.

**Comment by lu4p** [comment]
1. Files need to be somehow put in `/app/backend/data/cache/tools/run_code` e.g. `/app/backend/data/cache/tools/run_code/hello.txt`
2. This file is then accessible in the webui as `http://webui_url/cache/tools/run_code/hello.txt`
3. The tool can provide a message with a markdown link, something like:
```json
{
    "type": "message",
    "data": {"content": "[Download File](http://webui_url/cache/tools/run_code/hello.txt)"},  
}
```


ChatGPT generates code like this, where it outputs the generated filename to stdout:
```py
import uuid

# Generate a list of 1000 UUIDs
uuid_list = [str(uuid.uuid4()) for _ in range(1000)]

# Save to a text file
file_path = '/mnt/data/uuids.txt'
with open(file_path, 'w') as f:
    for uid in uuid_list:
        f.write(uid + '\n')

file_path
```

**Comment by EtiennePerot** [comment]
Excellent. With all these in place, I think this is all feasible now.

To prevent abuse, the tool probably needs a few more valves:
- Maximum number of files that can be stored (per user)
- Maximum total size of files stored (per user)
- Maximum age of the files before they are deleted (presumably by future iterations of the tool)

**Comment by EtiennePerot** [comment]
Another related idea: Letting the sandboxed code have a persistent per-chat "session" directory, in which files created in one session are available in future executions of the tool within the same chat. I'm not sure the current crop of models are really smart enough to think about persistence in this way, but good to think about in advance.

**Comment by EtiennePerot** [comment]
This is now implemented as of 44b33e38cd9ed43306697a4b22047ea9293ff112. Enjoy!

![Code execution function](https://github.com/EtiennePerot/open-webui-code-execution/blob/master/res/code-execution-function.gif?raw=true)

**Comment by sultanjulyan** [comment]
I have tried it and succeeded, thank you all.
